File: Greedy/Silver/P2_2015-High_Card_Wins/Silver_P2_2015-High_Card_Wins.py
with open("highcard.in") as f:
    n = int(f.readline())
    Elsie = [int(f.readline()) for _ in range(n)]
Elsie.sort()  # Sort Elsie's cards
Bessie = sorted(set(range(1, 2 * n + 1)) - set(Elsie)) # Calculate the cards Bessie has

l = r = 0
while r < n:
    # Find the first card in Bessie's set that is larger than Elsie's current card
    while r < n and Elsie[l] > Bessie[r]:
        r += 1
    # Move to the next round
    l += 1
    r += 1

print(l, file=open("highcard.out", "w"))

File: Greedy/Silver/P2_2015-High_Card_Wins/Silver_P2_2015-High_Card_Wins.cpp
#include <bits/stdc++.h>
using namespace std;

int n, Bessie[50001], Elsie[50002];

int main() {
    freopen("highcard.in", "r", stdin);
    freopen("highcard.out", "w", stdout);
    cin >> n; 
    for (int i = 1; i <= n; i++) 
        cin >> Elsie[i];  
    Elsie[n + 1] = n * 2;  // Add a boundary value to prevent bessie cards from missing statistics
    sort(Elsie + 1, Elsie + 1 + n);  // Sort Elsie's cards

    // Calculate the cards Bessie has
    for (int i = 1, j = 1; i <= n + 1; i++) 
        for (int k = Elsie[i - 1] + 1; k < Elsie[i]; k++)
            Bessie[j++] = k;  // Fill Bessie's cards with the remaining cards not in Elsie's set

    int l = 1, r = 1;
    while (r <= n) {
        // Find the first card in Bessie's set that is larger than Elsie's current card
        while (r < n && Elsie[l] > Bessie[r])
            r++;
        l++, r++;  // Move to the next round
    }
    cout << l - 1 << endl;
}
